### PR TITLE
feat: add `align_chunking` option to preserve deterministic chunk boundaries across workers

### LINKED ISSUE #758: SAME optimized dataset for DIFFERENT num_workers
## 🚀 Feature

Current the optimized dataset seems different if I'm using different `num_workers` chunk-$worker-$chunkid, can I make the final optimized always the same for different num_workers? Or can I reorganize the optimized dataset?

### PR DESCRIPTION
**Pull request by deependujha**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lit-data/blob/main/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #758

This PR introduces a new `align_chunking` option to `optimize()` and internal item-mapping logic. When enabled, items are assigned to workers using a deterministic chunk-aligned strategy that mirrors the single-worker chunk layout. All workers receive only full chunks of the specified size, except for the final worker, which may receive a partial chunk.

This eliminates the inconsistent partial-chunk patterns that occur when multiple workers independently chunk their own subsets, ensuring downstream samplers and loaders see stable chunk boundaries regardless of `num_workers`.

### Key points:

• Supports both `None` (current behavior) and an integer chunk size.
• Guarantees chunk alignment equivalent to `num_workers=1` & `num_nodes = 1`.
• Multi-node support ✅.
• Adds thorough tests verifying alignment behavior.

This provides a predictable chunk layout for users who rely on `chunk_size` semantics, especially when batch size is a multiple of chunk size, **`while maintaining full backwards compatibility`**.

---

### Code

- Optimize dataset
```python
import litdata as ld
import torch


def getter(index: int):
    return torch.Tensor(torch.full((4096, 128), index, dtype=torch.long))


if __name__ == "__main__":
    ld.optimize(
        fn=getter,
        inputs=list(range(7 * 64)),
        chunk_size=64,
        align_chunking=True,
        output_dir="worker_size_1",
        num_workers=1,
        mode="overwrite",
    )
    ld.optimize(
        fn=getter,
        inputs=list(range(7 * 64)),
        align_chunking=True,
        chunk_size=64,
        output_dir="worker_size_2",
        num_workers=2,
        mode="overwrite",
    )
```

- Stream and assert that each batch is from same chunk only.
```python
import litdata as ld

if __name__ == "__main__":
    print(" Worker num = 1 ".center(60, "-"))
    dataset = ld.StreamingDataset("worker_size_1/", shuffle=True)
    dataloader = ld.StreamingDataLoader(dataset, batch_size=64, num_workers=1)
    for i, batch in enumerate(dataloader):
        index = batch[:, 0, 0]
        min_item = index.min().item()
        max_item = index.max().item()
        assert max_item - min_item < 64, "each batch should belong to the same chunk."
        print(f"{i} ~ max-min-diff:", (index.max() - index.min()).item())


    print(" Worker num = 2 ".center(60, "-"))
    dataset = ld.StreamingDataset("worker_size_2/", shuffle=True)
    dataloader = ld.StreamingDataLoader(dataset, batch_size=64, num_workers=1)
    for i, batch in enumerate(dataloader):
        index = batch[:, 0, 0]
        min_item = index.min().item()
        max_item = index.max().item()
        assert max_item - min_item < 64, "each batch should belong to the same chunk."
        print(f"{i} ~ max-min-diff:", (index.max() - index.min()).item())
```

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/litdata/processing/data_processor.py [hunk 1]
```diff
@@ -299,24 +299,53 @@ def _upload_fn(
...
+        num_items_per_worker = [chunks_per_worker * align_chunking for _ in range(world_size - 1)]
+
+        # Last worker receives all remaining items (full chunks + optional tail)
+        remaining = len(user_items) - sum(num_items_per_worker)
+        num_items_per_worker.append(remaining)
+
+    else:
+        num_items_per_worker = len(user_items) // world_size
+
+        num_items_per_worker: list[int] = [num_items_per_worker for _ in range(world_size)]
```
[the remark this answers is not in the document]
Comment (deependujha) [reply src/litdata/processing/data_processor.py]: no intention to change original code

**Review by pwgardipee — APPROVED** [review]

File: src/litdata/processing/data_processor.py [hunk 2]
```diff
@@ -299,24 +299,53 @@ def _upload_fn(
...
-    for worker_idx in range(len(num_items_per_worker) - 1, -1, -1):
-        if reminder == 0:
-            break
-        num_items_per_worker[worker_idx] += 1
-        reminder -= 1
+        # Compute how many full chunks each worker can take
+        full_chunks = len(user_items) // chunk_size
+        chunks_per_worker = full_chunks // world_size
+
+        # Assign full chunks to all workers except the last
```
Comment (bhimrazy) [inline src/litdata/processing/data_processor.py:340]: @deependujha , we have  a minor edge-case issue with worker assignment. 
In cases like:

For eg: 
```
import torch
import litdata as ld


def getter(index: int):
    return torch.Tensor(torch.full((4096, 128), index, dtype=torch.long))


if __name__ == "__main__":
    ld.optimize(
        fn=getter,
        inputs=list(range(100)),
        chunk_size=64,
        align_chunking=True,
        output_dir="test_idle_workers",
        num_workers=2,
    )
```

In this case: 
Worker 0: Gets 0 * 64 = 0 items.
Worker 1: Gets 100 - 0 = 100 items.

Although this may be a minor edge case, should we consider adding a warning when `chunks_per_worker == 0` to indicate that some workers will be idle?

File: src/litdata/processing/functions.py [hunk 3]
```diff
@@ -428,6 +429,9 @@ def optimize(
         weights: Provide an associated weight to each input. This is used to balance work among workers.
         chunk_size: The maximum number of elements to hold within a chunk.
         chunk_bytes: The maximum number of bytes to hold within a chunk.
+        align_chunking: Ensures chunk boundaries match the single-worker layout by packing full chunks first
+            and placing all remaining items in the final worker. Each worker will receive chunks of this size,
+            except possibly the last worker which may receive a smaller chunk.
```
Comment (bhimrazy) [inline src/litdata/processing/functions.py]:
```python
align_chunking: If True, items are assigned to workers to ensure that every chunk (except the last one in the entire dataset) 
    has exactly `chunk_size` items. This mimics the chunk layout of a single-worker run. 
    Warning: This may result in uneven workload distribution or idle workers if the data size isn't large enough 
    to saturate all workers with full chunks.
```

Not necessarily this wording, but maybe a related warning could be added here.

File: tests/processing/test_functions.py [hunk 4]
```diff
@@ -104,6 +104,10 @@ def compress(index):
     return index, index**2
 
 
+def getter(index: int):
+    return torch.Tensor(torch.full((4096, 128), index, dtype=torch.long))
```
Comment (bhimrazy) [inline tests/processing/test_functions.py]:
```suggestion
    return index
```
I think we can simply return `index` here, or use an existing helper function if there’s one available.

File: tests/processing/test_functions.py [hunk 5]
```diff
@@ -123,6 +127,54 @@ def random_image(index):
...
+            chunk_bytes="1MB",
+            output_dir=str(output_dir),
+            num_workers=1,
+            align_chunking=True,
+        )
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="too slow")
+@pytest.mark.parametrize(
+    ("num_workers", "expected_worker_chunks"),
```
Comment (bhimrazy) [inline tests/processing/test_functions.py]: @deependujha , I think it might be better to define the chunk size here instead of `expected_worker_chunks`, since we can calculate and verify. We could then verify this across a few chunk sizes and input sizes, as the behavior should hold for any combination, wdyt ?

File: tests/streaming/test_dataloader.py [hunk 6]
```diff
@@ -496,3 +497,33 @@ def test_dataloader_dataset_transform_inheritance(tmpdir, shuffle):
     # Verify that the transform is applied correctly
     for i, item in enumerate(complete_data):
         assert item == i * 2, f"Expected {i * 2}, got {item}"
+
+
+def getter(index: int):
+    return torch.Tensor(torch.full((4096, 128), index, dtype=torch.long))
```
Comment (bhimrazy) [inline tests/streaming/test_dataloader.py]:
```suggestion
    return torch.Tensor(torch.full((2, 2), index, dtype=torch.long))
```

Also, I think we can omit this test (i.e. `test_dataloader_with_align_chunking`), as it may not hold true for all cases. Its behavior seems to depend on the batch size, seed when shuffling is enabled. 
Alternatively, we might need to add a comment noting that the batch size should match the chunk size, or clarify the expected conditions.

File: src/litdata/processing/data_processor.py (hunk 2 above)
Comment (deependujha) [reply src/litdata/processing/data_processor.py:340]: sure, adding a warning makes sense. Also, I guess, it simply means, your dataset is too small.

File: tests/streaming/test_dataloader.py (hunk 6 above)
Comment (deependujha) [reply tests/streaming/test_dataloader.py]: adding a comment seems better option.

File: tests/processing/test_functions.py (hunk 5 above)
Comment (deependujha) [reply tests/processing/test_functions.py]: great idea. thanks.

File: tests/processing/test_functions.py (hunk 4 above)
Comment (deependujha) [reply tests/processing/test_functions.py]: thanks, i didn't give this much thought and simply went ahead with what was shared by user.

Comment (deependujha) [reply tests/processing/test_functions.py]: updated.
